cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.tephra.Transaction;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * HBase implementation of {@link MessageTable}.
+ */
+public class HBaseMessageTable implements MessageTable {
+  private static final byte[] PAYLOAD_COL = Bytes.toBytes('p');
+  private static final byte[] TX_COL = Bytes.toBytes('t');
+
+  private final HBaseTableUtil tableUtil;
+  private final byte[] columnFamily;
+  private final HTable hTable;
+
+  private long writeTimestamp;
+  private short seqId;
+
+  public HBaseMessageTable(HBaseTableUtil tableUtil, HTable hTable, byte[] columnFamily) {
+    this.tableUtil = tableUtil;
+    this.hTable = hTable;
+    this.columnFamily = Arrays.copyOf(columnFamily, columnFamily.length);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, long startTime, int limit, @Nullable Transaction transaction)
+    throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = new byte[topic.length + Long.BYTES];
+    Bytes.putBytes(startRow, 0, topic, 0, topic.length);
+    Bytes.putLong(startRow, topic.length, startTime);
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = hTable.getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive,
+                                        int limit, @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = new byte[topic.length + Long.BYTES + Short.BYTES];
+    Bytes.putBytes(startRow, 0, topic, 0, topic.length);
+    Bytes.putLong(startRow, topic.length, messageId.getPublishTimestamp());
+    Bytes.putShort(startRow, topic.length + Long.BYTES, messageId.getSequenceId());
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = hTable.getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    long writeTs = System.currentTimeMillis();
+    if (writeTs != writeTimestamp) {
+      seqId = 0;
+    }
+    writeTimestamp = writeTs;
+    List<Put> batchPuts = new ArrayList<>();
+    while (entries.hasNext()) {
+      Entry entry = entries.next();
+      byte[] topic = MessagingUtils.toBytes(entry.getTopicId());
```

Comment:
We can save a bit of new array creation (both the `topic` and `tableKey`) if the topic doesn't change (which is the case when a client support multiple payloads in one request).